src/tokens.ts

```diff
@@ -0,0 +1,16 @@
+import { Token } from '@lumino/coreutils';
+
+namespace SchedulerExtension {
+  export interface ICustomEnvironmentProps {
+    state: any;
+    setState: any;
+  }
```

Comment:
We should definitely make this type a bit more strict, and give developers a way of specifying the keys in their form extension. Let's say the type of the base form state is called `FormState`:

```
namespace SchedulerExtension {
  export type CustomFormState = Record<never, never>;

  export interface ICustomEnvironmentProps {
    state: FormState & Partial<CustomFormState>;
    setState: (state: FormState & Partial<CustomFormState>) => unknown;
  }
}
```

Developers should then be able to override `CustomFormState` via [declaration merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html#merging-namespaces). However, I have not tested this as of yet.

Reply:
We should not provide arbitrary props, this should conform to the CreateModel attributes as the parent form will be passing in these props. Also, the custom form has to serialize all the extra properties (anything that doesn't fit into the top level CreateModel attributes) into "runtime_environment_parameters".

```ts
namespace Scheduler {
    export type EnvironmentParameterValue = string | number | boolean

    export interface CreateJobFormState {
        jobName: string
        inputUri: string
        outputPrefix: string
        runtimeEnvironmentName: string
        runtimeEnvironmentParameters?: { [key: string]: EnvironmentParameterValue }
        isValid: boolean
        // other attributes from api model
    }

    export interface ICustomEnvironmentProps {
        mode: 'Create' | 'View'
        state: CreateJobFormState
        setState: (state: CreateJobFormState) => void
    }
}
``` 
A few discussion points:
- We are calling this custom form as 'Advanced Options' in the UI, should we align with that naming here?
- I haven't thought about the Job Definition form, and if the same extension point will be used. If that's the case, then the state will support either type.